cloudevents/sdk/http_events.py

```diff
@@ -11,183 +11,148 @@
 #    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 #    License for the specific language governing permissions and limitations
 #    under the License.
-import copy
-
-import io
 
+import copy
+import datetime
 import json
+import io
 import typing
+import uuid
 
 from cloudevents.sdk import converters
 from cloudevents.sdk import marshaller
+from cloudevents.sdk import types
+from cloudevents.sdk.converters import binary
+from cloudevents.sdk.converters import structured
 
 from cloudevents.sdk.event import v03, v1
 
 
 class CloudEvent():
     """
     Python-friendly cloudevent class supporting v1 events
-    Currently only supports binary content mode CloudEvents
+    Supports both binary and structured mode CloudEvents
     """
 
-    def __init__(
-            self,
-            data: typing.Union[dict, None],
-            headers: dict = {},
-            data_unmarshaller: typing.Callable = lambda x: x,
+    @classmethod
+    def FromHttp(
+        cls,
+        data: typing.Union[str, bytes],
+        headers: dict = {},
+        data_unmarshaller: types.UnmarshallerType = None
     ):
```

Comment:
Why do we decide to not follow the [pep8 convention for naming of python methods](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):
pep8 says they should be:
```python
    def from_http(...) -> ...:
        ...
```

Also can we make sure to do "complete type hints":
```suggestion
    def from_http(
        cls,
        data: typing.IO,
        headers: typing.Dict[str, typing.Any],
        data_unmarshaller: types.UnmarshallerType = None
    ) -> CloudEvent:
```

Reply:
Hmm -- it turns out that I can't use `-> CloudEvent` to document the return type, because the class is not defined until after all the functions have been defined (including this one).

Thoughts?

Reply:
> Hmm -- it turns out that I can't use `-> CloudEvent` to document the return type, because the class is not defined until after all the functions have been defined (including this one).
> 
> Thoughts?

for this add the following at the top:

Reply:
For this:
> typing.IO expects to have a read() method, but a quick survey of 3 popular Python libraries (flask, requests, sanic), suggests that the normal exposure of the request body is via a string where the library handles the read, rather than exposing the underlying socket.

I just used the exact same typehints that the current implementation is using here: https://github.com/cloudevents/sdk-python/blob/master/cloudevents/sdk/marshaller.py#L45

BUT after reading your comment, I think you are right!
And for the headers I also think you are right!

So, double 👍

Reply:
It turns out that [PEP 563](https://www.python.org/dev/peps/pep-0563/) is only supported in Python 3.7 or later.

Reply:
and that's exactly why the added this feature in the "future" module.

And this line in the imports:
```
from __future__ import annotations
```

And it should work.

Reply:
It doesn't work on Python 3.6, which is in the test suite.

I tried adding a PR to do that, and the 3.6 tests (only) failed. If we can abandon 3.6, I'd be happy to make the change.

Reply:
[Python 3.6 won't be EOL until Dec 2021](https://python-release-cycle.glitch.me/), so I think we should probably continue to support it.